test/unit/io/sequence_file/sequence_file_format_sam_test.cpp

```diff
@@ -0,0 +1,285 @@
+struct write : public ::testing::Test
+{
+    std::vector<dna5_vector> seqs
+    {
+        "ACGT"_dna5,
+        "AGGCTGNAGGCTGNAGGCTGNAGGCTGNAGGCTGNAGGCTGNAGGCTGNAGGCTGNAGGCTGNAGGCTGNAGGCTGNAGGCTGNAGGCTGN"_dna5,
+        "GGAGTATAATATATATATATAT"_dna5
+    };
+    std::vector<std::string> ids
+    {
+        "TEST 1",
+        "Test2",
+        "Test3"
+    };
+    std::vector<std::vector<phred42>> quals
+    {
+        { "!##$"_phred42 },
+        { "!##$&'()*+,-./+)*+,-)*+,-)*+,-)*+,BDEBDEBDEBDEBDEBDEBDEBDEBDEBDEBDEBDEBDEBDEBDBDDEBDBEEBEBE"_phred42 },
+        { "!!*+,-./+*+,-./+!!FF!!"_phred42 },
+    };
+    sequence_file_format_sam format;
+    sequence_file_output_options options;
+    std::ostringstream ostream;
+    void do_write_test()
+    {
+        for (unsigned i = 0; i < 3; ++i)
+            EXPECT_NO_THROW(( format.write(ostream, options, seqs[i], ids[i], quals[i]) ));
+        ostream.flush();
+    }
+};
+TEST_F(write, arg_handling_id_missing)
+{
+    EXPECT_THROW( (format.write(ostream, options, seqs[0], std::ignore, quals[0])),
+                   std::logic_error );
+}
+
+TEST_F(write, arg_handling_id_empty)
+{
+    EXPECT_THROW( (format.write(ostream, options, seqs[0], std::string_view{""}, quals[0])),
+                   std::runtime_error );
+}
+
+TEST_F(write, arg_handling_seq_missing)
+{
+    EXPECT_THROW( (format.write(ostream, options, std::ignore, ids[0], quals[0])),
+                   std::logic_error );
+}
+
+TEST_F(write, arg_handling_seq_empty)
+{
+    EXPECT_THROW( (format.write(ostream, options, std::string_view{""}, ids[0], quals[0])),
+                   std::runtime_error );
+}
+
+TEST_F(write, arg_handling_qual_missing)
+{
+    EXPECT_THROW( (format.write(ostream, options, seqs[0], ids[0], std::ignore)),
+                   std::logic_error );
+}
+
+TEST_F(write, arg_handling_qual_empty)
+{
+    EXPECT_THROW( (format.write(ostream, options, seqs[0], ids[0], std::string_view{""})),
+                   std::runtime_error );
+}
+TEST_F(write, default_options)
+{
+    std::string comp
+    {
+      "TEST 1\t0\t*\t0\t0\t*\t*\t0\t0\tACGT\t!##$\n"
+      "Test2\t0\t*\t0\t0\t*\t*\t0\t0\tAGGCTGNAGGCTGNAGGCTGNAGGCTGNAGGCTGNAGGCTGNAGGCTGNAGGCTGNAGGCTGNAGGCTGNAGGCTGNAGGC"
+      "TGNAGGCTGN\t!##$&'()*+,-./+)*+,-)*+,-)*+,-)*+,BDEBDEBDEBDEBDEBDEBDEBDEBDEBDEBDEBDEBDEBDEBDBDDEBDBEEBEBE\n"
+      "Test3\t0\t*\t0\t0\t*\t*\t0\t0\tGGAGTATAATATATATATATAT\t!!*+,-./+*+,-./+!!FF!!\n"
+    };
+
+    do_write_test();
+
+    EXPECT_EQ(ostream.str(), comp);
+}
+
+
+
+// No qualities given
+TEST_F(write, no_qual)
+{
+    std::string comp
+    {
+      "TEST 1\t0\t*\t0\t0\t*\t*\t0\t0\tACGT\t*\n"
+      "Test2\t0\t*\t0\t0\t*\t*\t0\t0\tAGGCTGNAGGCTGNAGGCTGNAGGCTGNAGGCTGNAGGCTGNAGGCTGNAGGCTGNAGGCTGNAGGCTGNAGGCTGNAGGC"
+      "TGNAGGCTGN\t*\n"
+      "Test3\t0\t*\t0\t0\t*\t*\t0\t0\tGGAGTATAATATATATATATAT\t*\n"
+    };
+      for (unsigned i = 0; i < 3; ++i)
+        quals[i] = { "*"_phred42 };
```

Comment:
This is cheating :wink:. The test should have a write function that does not take any qualities and then automatically writes `*` for the quality fields!